From 1d92e5a21012e4f2d353d3673afc57c86cc66fa6 Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Tue, 10 Jul 2018 21:44:44 +0800
Subject: [PATCH 4136/5242] MGS-4061 gpu: imx: dpu-blit: fix kernel panic in
 pm test

commit  c6c4eb39df261385e36781dc30715619909410d7 from
https://source.codeaurora.org/external/imx/linux-imx.git

resume will increase unlock counter, max allowed value is 15.
suspend need decrease unlock counter to avoid overflow panic.

Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu-blit/dpu-blit-registers.h |    2 ++
 drivers/gpu/imx/dpu-blit/dpu-blit.c           |    6 ++++++
 2 files changed, 8 insertions(+)

diff --git a/drivers/gpu/imx/dpu-blit/dpu-blit-registers.h b/drivers/gpu/imx/dpu-blit/dpu-blit-registers.h
index d4c28d1..3d86b27 100644
--- a/drivers/gpu/imx/dpu-blit/dpu-blit-registers.h
+++ b/drivers/gpu/imx/dpu-blit/dpu-blit-registers.h
@@ -250,9 +250,11 @@
 #define CMDSEQ_HIF                                  ((uint32_t)(0x400))
 
 #define CMDSEQ_LOCKUNLOCKHIF                        ((uint32_t)(0x500))
+#define CMDSEQ_LOCKUNLOCKHIF_LOCKUNLOCKHIF__LOCK_KEY 0x5651F763U
 #define CMDSEQ_LOCKUNLOCKHIF_LOCKUNLOCKHIF__UNLOCK_KEY 0x691DB936U
 
 #define CMDSEQ_LOCKUNLOCK                           ((uint32_t)(0x580))
+#define CMDSEQ_LOCKUNLOCK_LOCKUNLOCK__LOCK_KEY      0x5651F763U
 #define CMDSEQ_LOCKUNLOCK_LOCKUNLOCK__UNLOCK_KEY    0x691DB936U
 
 #define CMDSEQ_BUFFERADDRESS                        ((uint32_t)(0x588))
diff --git a/drivers/gpu/imx/dpu-blit/dpu-blit.c b/drivers/gpu/imx/dpu-blit/dpu-blit.c
index 620d532..bb747d4 100644
--- a/drivers/gpu/imx/dpu-blit/dpu-blit.c
+++ b/drivers/gpu/imx/dpu-blit/dpu-blit.c
@@ -422,6 +422,12 @@ int dpu_bliteng_init(struct dpu_bliteng *dpu_bliteng)
 
 void dpu_bliteng_fini(struct dpu_bliteng *dpu_bliteng)
 {
+	/* LockUnlock and LockUnlockHIF */
+	dpu_be_write(dpu_bliteng, CMDSEQ_LOCKUNLOCKHIF_LOCKUNLOCKHIF__LOCK_KEY,
+		CMDSEQ_LOCKUNLOCKHIF);
+	dpu_be_write(dpu_bliteng, CMDSEQ_LOCKUNLOCK_LOCKUNLOCK__LOCK_KEY,
+		CMDSEQ_LOCKUNLOCK);
+
 	kfree(dpu_bliteng->cmd_list);
 
 	if (dpu_bliteng->buffer_addr_virt)
-- 
1.7.9.5

